pyquil/quilatom.py

```diff
@@ -245,54 +261,57 @@ class Expression(object):
 
     This class overrides all the Python operators that are supported by Quil.
     """
-    def __str__(self):
+    def __str__(self) -> str:
         return _expression_to_string(self)
 
-    def __repr__(self):
+    def __repr__(self) -> str:
         return str(self.__class__.__name__) + '(' + ','.join(map(repr, self.__dict__.values())) + ')'
 
-    def __add__(self, other):
+    def __add__(self, other: ExpressionOrValue) -> 'Add':
         return Add(self, other)
 
-    def __radd__(self, other):
+    def __radd__(self, other: ExpressionOrValue) -> 'Add':
         return Add(other, self)
 
-    def __sub__(self, other):
+    def __sub__(self, other: ExpressionOrValue) -> 'Sub':
         return Sub(self, other)
 
-    def __rsub__(self, other):
+    def __rsub__(self, other: ExpressionOrValue) -> 'Sub':
         return Sub(other, self)
 
-    def __mul__(self, other):
+    def __mul__(self, other: ExpressionOrValue) -> 'Mul':
         return Mul(self, other)
 
-    def __rmul__(self, other):
+    def __rmul__(self, other: ExpressionOrValue) -> 'Mul':
         return Mul(other, self)
 
-    def __div__(self, other):
+    def __div__(self, other: ExpressionOrValue) -> 'Div':
         return Div(self, other)
 
     __truediv__ = __div__
 
-    def __rdiv__(self, other):
+    def __rdiv__(self, other: ExpressionOrValue) -> 'Div':
         return Div(other, self)
 
     __rtruediv__ = __rdiv__
 
-    def __pow__(self, other):
+    def __pow__(self, other: ExpressionOrValue) -> 'Pow':
         return Pow(self, other)
 
-    def __rpow__(self, other):
+    def __rpow__(self, other: ExpressionOrValue) -> 'Pow':
         return Pow(other, self)
 
-    def __neg__(self):
+    def __neg__(self) -> 'Mul':
         return Mul(-1, self)
 
-    def _substitute(self, d):
+    def _substitute(self, d: Any) -> 'ExpressionOrValue':
         return self
 
 
-def substitute(expr, d):
+ParameterSubstitutionsDict = Dict['Parameter', ExpressionValue]
```

Comment:
I'm attempting to follow the (maybe informal) naming convention I found elsewhere for these type aliases. For example, `pyquil.api._quantum_computer.Executable` is an alias for `Union[BinaryExecutableResponse, PyQuilExecutableResponse]`.

Maybe it would help to suffix type aliases like this with `Type` or `TypeAlias` something, to make it clear during auto-completion that these aren't full-fledged classes, but rather only useful for type hints. On the other hand, any such naming convention is bound to be at least occasionally detrimental to readability.

Reply:
For better or worse, all the type hints now have a `Designator` suffix attached. In some cases, this is stretching the meaning of a "designator" in the Common Lisp sense of the word, but in many of the cases it fits just fine and at least provides a uniform naming convention that makes it easier to tell these are type aliases and not Real Python Classes(TM).